lib/src/arithmetic/booth.dart

```diff
@@ -0,0 +1,744 @@
+      m[lastRow][i] =
+          lastAddend[i] ^ (i < align ? propagate[lastRow][i] : Const(0));
+    }
+
+    final remainders = List.filled(rows, Logic());
+    for (var row = 0; row < lastRow; row++) {
+      remainders[row] = propagate[row][shift - 1];
+    }
+    remainders[lastRow] = propagate[lastRow][align > 0 ? align : 0];
+
+    // Merge 'm' into the LSBs of each addend
+    for (var row = 0; row < rows; row++) {
+      final addend = partialProducts[row];
+      if (row > 0) {
+        final mLimit = (row == lastRow) ? align : shift - 1;
+        for (var i = 0; i < mLimit; i++) {
+          addend[i] = m[row][i];
+        }
+        // Stop bits
+        _addStopSignFlip(addend, ~signs[row]);
+        addend
+          ..insert(0, remainders[row - 1])
+          ..addAll(List.filled(shift - 1, Const(1)));
+        rowShift[row] -= 1;
+      } else {
+        // First row
+        for (var i = 0; i < shift - 1; i++) {
+          firstAddend[i] = m[0][i];
+        }
+      }
+    }
+
+    // Insert the lastRow sign:  Either in firstRow's Q if there is a
+    // collision or in another row if it lands beyond the Q sign extension
+
+    final firstSign = signed ? firstAddend.last : signs[0];
+    final lastSign = remainders[lastRow];
+    // Compute Sign extension MSBs for firstRow
+    final qLen = shift + 1;
+    final insertSignPos = (align > 0) ? 0 : -align;
+    final q = List.filled(min(qLen, insertSignPos), firstSign, growable: true);
+    if (insertSignPos < qLen) {
+      // At sign insertion position
+      q.add(firstSign ^ lastSign);
+      if (insertSignPos == qLen - 1) {
+        q[insertSignPos] = ~q[insertSignPos];
+        q.add(~(firstSign | q[insertSignPos]));
+      } else {
+        q
+          ..addAll(List.filled(qLen - insertSignPos - 2, firstSign & ~lastSign))
+          ..add(~(firstSign & ~lastSign));
+      }
+    }
+
+    if (-align >= q.length) {
+      q.last = ~firstSign;
+    }
+    _addStopSign(firstAddend, q[0]);
+    firstAddend.addAll(q.getRange(1, q.length));
+
+    if (-align >= q.length) {
+      final finalCarryRelPos =
+          lastRowSignPos - selector.width - shift + (signed ? 1 : 0);
+      final finalCarryRow = (finalCarryRelPos / shift).floor();
+      final curRowLength =
+          partialProducts[finalCarryRow].length + rowShift[finalCarryRow];
+
+      partialProducts[finalCarryRow]
+        ..addAll(List.filled(lastRowSignPos - curRowLength, Const(0)))
+        ..add(remainders[lastRow]);
+    }
+    if (shift == 1) {
+      lastAddend.add(Const(1));
+    }
+  }
+
+  /// Return the actual largest width of all rows
+  int maxWidth() {
+    var maxW = 0;
+    for (var row = 0; row < rows; row++) {
+      final entry = partialProducts[row];
+      if (entry.length + rowShift[row] > maxW) {
+        maxW = entry.length + rowShift[row];
+      }
+    }
+    return maxW;
+  }
+
+  /// Accumulate the partial products and return as BigInt
+  BigInt evaluate() {
```

Comment:
I found a really nice debugging ability was to be able to walk all structures and print out their current boolean values while in a test.  

I tried to create an extension, but I lose the ability to do a print() on the object (it won't allow me to define toString).  So I need to rethink.  This idea is definitely a test-time accessor to the classes, so an extension makes some sense.

Reply:
yeah it seems useful, just maybe not quite right where it lives (visible as public API).  you could always make a new function rather than override toString() in the extension

Reply:
I noticed that you moved it into an extension, but the extension is exported as part of the rohd_hcl lib, so it's still visible externally